stdlib/LinearAlgebra/src/eigen.jl

```diff
@@ -647,9 +647,34 @@ function Base.isequal(A::Eigen, B::Eigen)
 end
 
 # Conversion methods
-
 ## Can we determine the source/result is Real?  This is not stored in the type Eigen
-AbstractMatrix(F::Eigen) = F.vectors * Diagonal(F.values) / F.vectors
+AbstractMatrix(F::Eigen) = F.are_vectors_unitary ?
+                             (F.vectors * Diagonal(F.values) * (F.vectors')) :
+                             (F.vectors * Diagonal(F.values) / F.vectors)
 AbstractArray(F::Eigen) = AbstractMatrix(F)
 Matrix(F::Eigen) = Array(AbstractArray(F))
 Array(F::Eigen) = Matrix(F)
+
+# Operations on Eigen
+# Only operations that can be efficiently implemented or conserve the eigen type are implemented
+adjoint(E::Eigen) = Eigen(adjoint.(E.values),
+                          copy(adjoint(E.are_vectors_unitary ? (E.vectors') : inv(E.vectors))),
+                          E.are_vectors_unitary)
+transpose(E::Eigen) = Eigen(transpose.(E.values),
+                            copy(transpose(E.are_vectors_unitary ? (E.vectors') : inv(E.vectors))),
+                            E.are_vectors_unitary && (eltype(E) <: Real))
+
+det(A::Eigen) = prod(A.values)
+tr(E::Eigen) = sum(E.values)
+
+*(E::Eigen, x::AbstractVector) = E.vectors * (E.values .* (E.are_vectors_unitary ? (E.vectors'x) : (E.vectors \ x)))
```

Comment:
That means a `O(n^3)` operation for each call. It would make sense to store `inv(vectors)'` in the `Eigen` structure here ( and remove `are_vectors_unitary`).

Reply:
Yes but this pays the cost upfront, even if the user doesn't need to do anything except computing the values/vectors, right?

Reply:
Oh, you mean in this function, compute the inverse of the vectors, and mutate E? I don't think we should do something as sophisticated as this, and having Eigen be mutable seems like begging for trouble. These are mostly convenience methods for playing around at the REPL or quick scripts, they are not meant to be used for performance-sensitive work, so over-engineering this adds complexity for not much gain.